### PR TITLE
Charlesmchen/refine attachment approval

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### TBD: exactly how large to make the approval preview.

screen shot 2017-10-23 at 12 09 58 pm
screen shot 2017-10-23 at 12 00 37 pm
screen shot 2017-10-23 at 12 38 12 pm

### CODE REVIEW

File: Signal/src/ViewControllers/ConversationView/ConversationInputToolbar.m [hunk 1]
```diff
@@ -214,12 +210,51 @@ - (void)ensureContentConstraints
...
+        + self.inputTextView.contentInset.bottom);
+    const CGFloat textViewHeight = ceil(MAX(kMinTextViewHeight, MIN(kMaxTextViewHeight, textViewDesiredHeight)));
+    const CGFloat kMinContentHeight = kMinTextViewHeight + textViewVInset * 2;
+
     if (self.attachmentToApprove) {
-        self.contentView.hidden = YES;
-        self.attachmentApprovalView.hidden = NO;
+        OWSAssert(self.attachmentView);
+
+        self.leftButtonWrapper.hidden = YES;
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationInputToolbar.m]: There are important changes here and below around making sure that subviews not relevant to the current mode are hidden and therefore won't receive input events.

File: Signal/src/ViewControllers/MediaMessageView.swift [hunk 2]
```diff
@@ -365,4 +390,51 @@ class MediaMessageView: UIView, OWSAudioAttachmentPlayerDelegate {
...
+
+    private func moviePlayerDidExitFullscreen() {
+        clearVideoPlayer()
+    }
+
+    private func clearVideoPlayer() {
+        videoPlayer?.stop()
+        videoPlayer?.view.removeFromSuperview()
+        videoPlayer = nil
+        ViewControllerUtils.setAudioIgnoresHardwareMuteSwitch(false)
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/MediaMessageView.swift]: I forgot what this method did and just took a look. It strikes me as weird that we'd be requesting a "record" session at this place. I know this is existing behavior, but I've made a card to audit the usage of this method.

File: Signal/src/ViewControllers/ConversationView/ConversationInputToolbar.m [hunk 3]
```diff
@@ -214,12 +210,51 @@ - (void)ensureContentConstraints
...
+    const int contentHSpacing = 6;
+
+    // We want to grow the text input area to fit its content within reason.
+    const CGFloat kMinTextViewHeight = ceil(self.inputTextView.font.lineHeight
+        + self.inputTextView.textContainerInset.top + self.inputTextView.textContainerInset.bottom
+        + self.inputTextView.contentInset.top + self.inputTextView.contentInset.bottom);
+    const CGFloat kMaxTextViewHeight = 100.f;
+    const CGFloat textViewDesiredHeight = (self.inputTextView.contentSize.height + self.inputTextView.contentInset.top
+        + self.inputTextView.contentInset.bottom);
+    const CGFloat textViewHeight = ceil(MAX(kMinTextViewHeight, MIN(kMaxTextViewHeight, textViewDesiredHeight)));
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/ViewControllers/ConversationView/ConversationInputToolbar.m]: Done.

File: Signal/src/ViewControllers/ConversationView/ConversationInputToolbar.m [hunk 4]
```diff
@@ -316,7 +338,7 @@ - (void)ensureContentConstraints
 
 - (void)ensureShouldShowVoiceMemoButton
 {
-    self.shouldShowVoiceMemoButton = self.inputTextView.trimmedText.length < 1;
+    self.shouldShowVoiceMemoButton = (self.attachmentToApprove != nil && self.inputTextView.trimmedText.length < 1);
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/ViewControllers/ConversationView/ConversationInputToolbar.m]: Good catch.  It's already fixed in the next PR.

File: Signal/src/ViewControllers/MediaMessageView.swift [hunk 5]
```diff
@@ -154,48 +160,67 @@ class MediaMessageView: UIView, OWSAudioAttachmentPlayerDelegate {
...
+        addSubviewWithScaleAspectFitLayout(view:animatedImageView, aspectRatio:aspectRatio)
+        contentView = animatedImageView
+    }
+
+    private func addSubviewWithScaleAspectFitLayout(view: UIView, aspectRatio: CGFloat) {
+        self.addSubview(view)
+        view.autoCenterInSuperview()
+        view.autoPin(toAspectRatio:aspectRatio)
+        view.autoMatch(.width, to: .width, of: self, withMultiplier: 1.0, relation: .lessThanOrEqual)
+        view.autoMatch(.height, to: .height, of: self, withMultiplier: 1.0, relation: .lessThanOrEqual)
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/ViewControllers/MediaMessageView.swift]: I've added a comment.

File: Signal/src/ViewControllers/MediaMessageView.swift (hunk 2 above)
Comment (charlesmchen) [reply Signal/src/ViewControllers/MediaMessageView.swift]: Good call.  I can't recall if this is a mistake or not.

**Pull request merged**